internal/db/schema/migrations/postgres/16/03_wh_credential_dimension.up.sql

```diff
@@ -0,0 +1,170 @@
+             project_id,            project_name,            project_description,
+             organization_id,       organization_name,       organization_description,
+             'Current',             current_timestamp,       'infinity'::timestamptz
+        from whx_credential_dimension_source
+       where credential_library_id = p_library_id
+         and session_id            = p_session_id
+         and target_id             = t_id
+         and credential_purpose    = p_credential_purpose
+      returning * into new_row;
+
+      return new_row.key;
+    end if;
+
+    return target.key;
+  end
+  $$ language plpgsql;
+
+  -- Run wh_upsert_credential_dimension for session_credential_dynamic row that is inserted.
+  create function wh_insert_session_credential_dynamic()
+    returns trigger
+  as $$
+  begin
+    perform wh_upsert_credential_dimension(new.session_id, new.library_id, new.credential_purpose);
+    return null;
+  end;
+  $$ language plpgsql;
+  create trigger wh_insert_session_credential_dynamic
+    after insert on session_credential_dynamic
+    for each row
+    execute function wh_insert_session_credential_dynamic();
+
+  -- wh_upsert_credentail_group determines if a new wh_credential_group needs to be
+  -- created due to changes to the coresponding wh_credential_dimensions. It then
+  -- updates the wh_session_accumulating_fact to associate it with the correct wh_credential_group.
+  create function wh_upsert_credentail_group()
+    returns trigger
+  as $$
+  declare
+    cg_key wh_dim_key;
+    t_id   wt_public_id;
+    s_id   wt_public_id;
+    c_key  wh_dim_key;
+  begin
+    select distinct scd.session_id into strict s_id
+      from new_table as scd;
+
+    select distinct s.target_id into strict t_id
+           from new_table as scd
+      left join session   as s   on s.public_id = scd.session_id;
+
+    -- based on query written by Michele Gaffney
+    with
+    credential_list (key) as (
+      select key
+        from wh_credential_dimension
+       where target_id = t_id
+         and credential_library_id in (select credential_library_id from new_table)
+    )
+    select distinct credential_group_key into cg_key
+      from wh_credential_group_membership a
+     where a.credential_key in (select key from credential_list)
+       and (select count(key) from credential_list) =
+           (
+            select count(b.credential_key)
+              from wh_credential_group_membership b
+             where a.credential_key = b.credential_key
+               and b.credential_key in (select key from credential_list)
+           )
+       and not exists
+           (
+            select 1
+              from wh_credential_group_membership b
+             where a.credential_key = b.credential_key
+               and b.credential_key not in (select key from credential_list)
+           )
```

Comment:
Does this where statement using this subquery do anything? We checked on line 121 that a.credential_key is in credential_list.  Since b.credential_key = a.credential_key this where clause can be reduced to confirming that a record doesn't exist where false, no?

Reply:
Maybe this should be 
```
       and not exists
           (
            select 1
              from wh_credential_group_membership b
             where a.credential_group_key = b.credential_group_key
               and b.credential_key not in (select key from credential_list)
           )
```